### ISSUE TITLE
Enhance stats

### ISSUE THREAD

**Issue by MarcoBuster** [tdlight-team/tdlight-telegram-bot-api#17]
Statistics can be helpful in many ways such as monitoring the instance and implement load balancers. 
- Currently, the stats format is easy to read for an human but hard to parse for a script so we need to add an option to retrieve them as JSON (with backwards compatibility). 
- @andrew-ld suggests to implement a public method to read (own) bot statistics from the public endpoint, like this: `https://botapi.example.org/botID:TOKEN/getStats` with a command line option or environment option to enable/disable it.

### Progress tracker
- [ ] JSON format
- [ ] Toggle to hide sensible data
- [ ] Per-bot public statistics
- [ ] Documentation

**Comment by iiiiii1wepfj** [comment]
@MarcoBuster 2 it's my idea

**Comment by luckydonald** [comment]
I'm currently working on 1, albeit using the usual `JsonBuilder` to get the stats.
I'm also adding a toggle to hide the sensible data (bot token so far) from it, so it can be served public (json or not).

**Comment by luckydonald** [comment]
I'm currently working on 1, albeit using the usual `JsonBuilder` to get the stats.
That way we can easily extract and re-use that for any bot method.

I'm also adding a toggle to hide the sensible data (bot token so far) from it, so it can be served public (json or not).

**Comment by luckydonald** [comment]
<img width="1300" alt="image" src="https://user-images.githubusercontent.com/2737108/101855546-dec22a80-3b63-11eb-82c3-5b282b9924c0.png">

**Comment by luckydonald** [comment]
Hmmm, I'm stuck with the bots.

It starts at 
https://github.com/tdlight-team/tdlight-telegram-bot-api/blob/14bf658b86d18e941a1338f1364f2b527ea4f52b/telegram-bot-api/ClientManager.cpp#L223-L229
goes through all the bots connected (which could be a very long list)
and writes as many as possible to the buffer, until that buffer is full:
https://github.com/tdlight-team/tdlight-telegram-bot-api/blob/14bf658b86d18e941a1338f1364f2b527ea4f52b/telegram-bot-api/ClientManager.cpp#L247-L250
However I haven't find a way to really mimic that with how the Json works.

**Comment by luckydonald** [comment]
I can probably have that write to the buffer as well, similar to how the text does it,
but in case of that overflow we would end up without the closing tags (`]}`) which is rather suboptimal.

I would say for a first implementation that probably wouldn't be an issue for most, as most of those self hosted instances will run with only a few bots, and if it get's too long you'll just have to fall back to the the non-json version for now.

**Comment by MarcoBuster** [comment]
Awesome work, thank you. Keep us updated!

**Comment by luckydonald** [comment]
Well I'm less in control of that buffer than I thought I would be when using the regular json stuff.
I think I will reduce the `"bots"` array to just a list of id's which then could be queried independently, over the normal api or something

**Comment by luckydonald** [comment]
While add it I also added a CLI flag to disable the display of sensible data (bot tokens + Webhook urls), which enables the stats to be hosted public.

**Comment by luckydonald** [comment]
If someone already wants to give it a try, 374928c3ab46eb5a779a1919d37ea7e753b3da6a seems to be a working one.
There are still a few fields missing.
Regarding the buffer size problem, I now show only the internal id + the ranking if it's more than 100 active bots.
